packages/protocol/contracts/layer1/preconf/impl/PreconfSlasher.sol

```diff
@@ -97,104 +103,68 @@ contract PreconfSlasher is IPreconfSlasher, EssentialContract {
     {
         EvidenceInvalidPreconfirmation memory evidence =
             abi.decode(_evidenceData, (EvidenceInvalidPreconfirmation));
-        require(
-            keccak256(evidence.preconfedBlockHeader.encodeRLP()) == _payload.blockHash,
-            InvalidBlockHeader()
-        );
-
-        ITaikoInbox.Batch memory batch = taikoInbox.v4GetBatch(uint64(_payload.batchId));
-        ITaikoInbox.TransitionState memory transition =
-            taikoInbox.v4GetBatchVerifyingTransition(uint64(_payload.batchId));
+        require(evidence.preconfedBlockHeader.hash() == _payload.blockHash, InvalidBlockHeader());
 
-        // Validate that the batch has been verified
-        require(transition.blockHash != bytes32(0), BatchNotVerified());
-
-        // Validate the pre-images in the evidence
-        require(
-            keccak256(abi.encode(evidence.batchMetadata)) == batch.metaHash, InvalidBatchMetadata()
-        );
-        require(
-            keccak256(abi.encode(evidence.batchInfo)) == evidence.batchMetadata.infoHash,
-            InvalidBatchInfo()
-        );
+        _verifyBatchData(_payload.batchId, evidence.batchMetadata, evidence.batchInfo);
 
         // Slash if the height of anchor block on the commitment is different from the
         // height of anchor block on the proposed block
         if (_payload.anchorId != evidence.batchInfo.anchorBlockId) {
             return getSlashAmount().invalidPreconf;
         }
 
+        // Ensure that the anchor block has not been reorged out
+        require(
+            _payload.anchorHash == evidence.batchInfo.anchorBlockHash, PossibleReorgOfAnchorBlock()
+        );
+
         // Check for reorgs if the committer missed the proposal
         if (evidence.batchInfo.proposer != _committer) {
             (bool success,) = LibPreconfConstants.getBeaconBlockRootContract().staticcall(
                 abi.encode(_payload.preconferSlotTimestamp)
             );
-
             // If the beacon block root is not available, it means that the preconfirmed block
             // was reorged out due to an L1 reorg.
             if (!success) {
                 return getSlashAmount().reorgedPreconf;
             }
         }
 
-        // Ensure that the anchor block has not been reorged out
-        require(
-            _payload.anchorHash == evidence.batchInfo.anchorBlockHash, PossibleReorgOfAnchorBlock()
-        );
+        // The preconfirmed blockhash must not match the hash of the proposed block for a
+        // preconfirmation violation
+        bytes32 l2BlockHash = evidence.blockhashProofs.l2BlockHeader.hash();
+        require(_payload.blockHash != l2BlockHash, PreconfirmationIsValid());
+
+        // Validate that the batch has been verified
+        ITaikoInbox.TransitionState memory transition =
```

Comment:
Is it a good idea to split this function into 2: one for slashing before the batch is verified, the other is for after the batch is verified...